src/sentry/release_health/metrics.py

```diff
@@ -606,3 +601,135 @@ def extract_row_info(row: Mapping[str, Union[OrganizationId, str]]) -> ReleaseNa
+        try:
+            release_key = tag_key(org_id, "release")
+            release_value = tag_value(org_id, release)
+            environment_key = tag_key(org_id, "environment")
+            status_key = tag_key(org_id, "session.status")
+        except MetricIndexNotFound:
+            # No need to query snuba if any of these is missing
+            return generate_defaults
+
+        environment_values = None
+        if environments is not None:
+            environment_values = get_tag_values_list(org_id, environments)
+
+        if environment_values == []:
+            # No need to query snuba with an empty list
+            return generate_defaults
+
+        status_init = tag_value(org_id, "init")
+        status_crashed = tag_value(org_id, "crashed")
+
+        conditions = [
+            Condition(Column("org_id"), Op.EQ, org_id),
+            Condition(Column("project_id"), Op.EQ, project_id),
+            Condition(Column(release_key), Op.EQ, release_value),
+            Condition(Column("timestamp"), Op.GTE, start),
+            Condition(Column(status_key), Op.IN, [status_init, status_crashed]),
+        ]
+        if environment_values is not None:
+            conditions.append(Condition(Column(environment_key), Op.IN, environment_values))
+
+        def query_stats(end: datetime) -> CrashFreeBreakdown:
+            def _get_data(entity_key: EntityKey, metric_name: str) -> Tuple[int, int]:
+                total = 0
+                crashed = 0
+                metric_id = try_get_string_index(org_id, metric_name)
+                if metric_id is not None:
+                    where = conditions + [
+                        Condition(Column("metric_id"), Op.EQ, metric_id),
+                        Condition(Column("timestamp"), Op.LT, end),
+                    ]
+                    data = raw_snql_query(
+                        Query(
+                            dataset=Dataset.Metrics.value,
+                            match=Entity(entity_key.value),
+                            select=[Column("value")],
+                            where=where,
+                            groupby=[Column(status_key)],
+                        ),
+                        referrer="release_health.metrics.crash-free-breakdown.session",
+                    )["data"]
+                    for row in data:
+                        if row[status_key] == status_init:
+                            total = int(row["value"])
+                        elif row[status_key] == status_crashed:
+                            crashed = int(row["value"])
+
+                return total, crashed
+
+            sessions_total, sessions_crashed = _get_data(EntityKey.MetricsCounters, "session")
+            users_total, users_crashed = _get_data(EntityKey.MetricsSets, "user")
+
+            return {
+                "date": end,
+                "total_users": users_total,
+                "crash_free_users": 100 - users_crashed / float(users_total) * 100
```

Comment:
I would unify how this is calculated everywhere because if i recall correctly @untitaker might have been doing this differently?

Reply:
No I did exactly this... we're just porting the formulas from sessions.py where it was also duplicated but also consistent.

Reply:
Let's dedup when we do https://getsentry.atlassian.net/browse/INGEST-456.